lib/freyja/resource_factory.rb

```diff
@@ -65,46 +65,36 @@ def perform(object)
         resource_factory = Hyrax.query_service.services.first.instance_variable_get(:@resource_factory)
 
         resource = ::Valkyrie::Persistence::Postgres::ORMConverter.new(object, resource_factory:).convert!
-        migrate_thumbnail!(resource:)
+
+        migrate_derivatives!(resource:)
         migrate_files!(resource:)
       end
 
       private
 
-      ##
-      # @param path [String] path to the expected thumbnail
-      #
-      # @return [TrueClass] when the thumbnail at the given path has not been
-      #         moved to the Valkyrie storage adapter.
-      # @return [FalseClass] when the thumbnail has been moved to the Valkyrie
-      #         storage adapter.
-      # @see #move_thumbnail_to_backup
-      def thumbnail_exists?(path)
-        path.present? && File.exist?(path)
-      end
-
-      def migrate_thumbnail!(resource:)
-        thumbnail_path = Hyrax::DerivativePath.derivative_path_for_reference(resource, 'thumbnail')
-        return unless thumbnail_exists?(thumbnail_path)
-
-        tempfile = Tempfile.new
-        tempfile.binmode
-        tempfile.write(File.read(thumbnail_path))
-
-        # NOTE: There are published events that may or may not be appropriate
-        # for this to call.  It's hard to know, given that ActiveFedora's
-        # thumbnail was never a "File" on a FileSet but was a unique creature.
-        # With Valkyrie that changes and we have a right and proper
-        # "Hyrax::PCDM::File" for the thumbnail.
-        Hyrax::ValkyrieUpload.file(
-          filename: resource.label,
-          file_set: resource,
-          io: tempfile,
-          use: Hyrax::FileMetadata::Use::THUMBNAIL_IMAGE,
-          user: User.find_or_initialize_by(User.user_key_field => resource.depositor)
-        )
-
-        move_thumbnail_to_backup(thumbnail_path)
+      def migrate_derivatives!(resource:)
+        member_ids = resource.member_ids
+        members = Hyrax.query_service.find_many_by_ids(ids: member_ids)
+
+        members.each do |object|
+          # @todo should we trigger a job if the member is a child work?
+          next unless object.is_a?(FileSet) || object.is_a?(Hyrax::FileSet)
+
+          paths = Hyrax::DerivativePath.derivatives_for_reference(object)
+          paths.each do |path|
+            next unless path.present?
+            path.each_child do |file|
+              file_path = path + '/' + file
+              content = File.read(file_path)
+              container = container_for(file)
+              mime_type = Marcel::MimeType.for(extension: File.extname(file))
+              directives = { url: file_path, container: container, mime_type: mime_type }
+              Hyrax::ValkyriePersistDerivatives.call(content, directives)
```

Comment:
ValkyriePersistDerivatives will do its own read of the file io, and with a passed block will auto close the file.
```suggestion
              File.open(file_path, 'rb') do |content|
                container = container_for(file)
                mime_type = Marcel::MimeType.for(extension: File.extname(file))
                directives = { url: file_path, container: container, mime_type: mime_type }
                Hyrax::ValkyriePersistDerivatives.call(content, directives)
              end
```